### ISSUE TITLE
woocommerce_rest_invalid_order_refund_id

### ISSUE THREAD

**Issue by luiscoutinh** [woocommerce/woocommerce#13392]
Hi,

I'm trying to get the refunds to a order. However, when I try to get the refunds of the order id 578 I get the refunds and some errors. The error reported is: woocommerce_rest_invalid_order_refund_id.

I'm using the lastest version of WooCommerce and this is a clean Wordpress installation that only have the WooCommerce plugin. I have created the order and saved it. After, I created some refunds of the same item.

This happens with other orders. And the number of errors is variable.

I've searched in API and this errors is only dispatched when the order doesn't exists on wp-content/plugins/woocommerce/includes/api/class-wc-rest-order-refunds-controller.php in prepare_item_for_response method.

To solve this problem temporarily, I'm filtering the results like:
```
return array_filter($results, function($refund) {
            return !isset($refund['errors']);
        });
```

**Above you can see an example of what I expect to get:**
```
array(3) {
  [0]=>
  array(6) {
    ["id"]=>
    int(581)
    ["date_created"]=>
    string(19) "2017-02-28T00:44:40"
    ["amount"]=>
    string(5) "29.34"
    ["reason"]=>
    string(0) ""
    ["line_items"]=>
    array(1) {
      [0]=>
      array(14) {
        ["id"]=>
        int(29)
        ["name"]=>
        string(13) "Woo Single #1"
        ["sku"]=>
        string(10) "woosingle1"
        ["product_id"]=>
        int(93)
        ["variation_id"]=>
        int(0)
        ["quantity"]=>
        int(-3)
        ["tax_class"]=>
        string(0) ""
        ["price"]=>
        string(6) "-23.85"
        ["subtotal"]=>
        string(6) "-23.85"
        ["subtotal_tax"]=>
        string(5) "-5.49"
        ["total"]=>
        string(6) "-23.85"
        ["total_tax"]=>
        string(5) "-5.49"
        ["taxes"]=>
        array(1) {
          [0]=>
          array(3) {
            ["id"]=>
            int(1)
            ["total"]=>
            string(7) "-5.4855"
            ["subtotal"]=>
            string(7) "-5.4855"
          }
        }
        ["meta"]=>
        array(0) {
        }
      }
    }
    ["_links"]=>
    array(3) {
      ["self"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(68) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds/581"
        }
      }
      ["collection"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(64) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds"
        }
      }
      ["up"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(56) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578"
        }
      }
    }
  }
  [1]=>
  array(6) {
    ["id"]=>
    int(580)
    ["date_created"]=>
    string(19) "2017-02-28T00:44:26"
    ["amount"]=>
    string(5) "19.56"
    ["reason"]=>
    string(0) ""
    ["line_items"]=>
    array(1) {
      [0]=>
      array(14) {
        ["id"]=>
        int(27)
        ["name"]=>
        string(13) "Woo Single #1"
        ["sku"]=>
        string(10) "woosingle1"
        ["product_id"]=>
        int(93)
        ["variation_id"]=>
        int(0)
        ["quantity"]=>
        int(-2)
        ["tax_class"]=>
        string(0) ""
        ["price"]=>
        string(6) "-15.90"
        ["subtotal"]=>
        string(6) "-15.90"
        ["subtotal_tax"]=>
        string(5) "-3.66"
        ["total"]=>
        string(6) "-15.90"
        ["total_tax"]=>
        string(5) "-3.66"
        ["taxes"]=>
        array(1) {
          [0]=>
          array(3) {
            ["id"]=>
            int(1)
            ["total"]=>
            string(6) "-3.657"
            ["subtotal"]=>
            string(6) "-3.657"
          }
        }
        ["meta"]=>
        array(0) {
        }
      }
    }
    ["_links"]=>
    array(3) {
      ["self"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(68) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds/580"
        }
      }
      ["collection"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(64) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds"
        }
      }
      ["up"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(56) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578"
        }
      }
    }
  }
  [2]=>
  array(6) {
    ["id"]=>
    int(579)
    ["date_created"]=>
    string(19) "2017-02-28T00:44:14"
    ["amount"]=>
    string(4) "9.78"
    ["reason"]=>
    string(0) ""
    ["line_items"]=>
    array(1) {
      [0]=>
      array(14) {
        ["id"]=>
        int(25)
        ["name"]=>
        string(13) "Woo Single #1"
        ["sku"]=>
        string(10) "woosingle1"
        ["product_id"]=>
        int(93)
        ["variation_id"]=>
        int(0)
        ["quantity"]=>
        int(-1)
        ["tax_class"]=>
        string(0) ""
        ["price"]=>
        string(5) "-7.95"
        ["subtotal"]=>
        string(5) "-7.95"
        ["subtotal_tax"]=>
        string(5) "-1.83"
        ["total"]=>
        string(5) "-7.95"
        ["total_tax"]=>
        string(5) "-1.83"
        ["taxes"]=>
        array(1) {
          [0]=>
          array(3) {
            ["id"]=>
            int(1)
            ["total"]=>
            string(7) "-1.8285"
            ["subtotal"]=>
            string(7) "-1.8285"
          }
        }
        ["meta"]=>
        array(0) {
        }
      }
    }
    ["_links"]=>
    array(3) {
      ["self"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(68) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds/579"
        }
      }
      ["collection"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(64) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds"
        }
      }
      ["up"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(56) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578"
        }
      }
    }
  }
}
```

**However, I'm getting this:**
```
array(7) {
  [0]=>
  array(6) {
    ["id"]=>
    int(581)
    ["date_created"]=>
    string(19) "2017-02-28T00:44:40"
    ["amount"]=>
    string(5) "29.34"
    ["reason"]=>
    string(0) ""
    ["line_items"]=>
    array(1) {
      [0]=>
      array(14) {
        ["id"]=>
        int(29)
        ["name"]=>
        string(13) "Woo Single #1"
        ["sku"]=>
        string(10) "woosingle1"
        ["product_id"]=>
        int(93)
        ["variation_id"]=>
        int(0)
        ["quantity"]=>
        int(-3)
        ["tax_class"]=>
        string(0) ""
        ["price"]=>
        string(6) "-23.85"
        ["subtotal"]=>
        string(6) "-23.85"
        ["subtotal_tax"]=>
        string(5) "-5.49"
        ["total"]=>
        string(6) "-23.85"
        ["total_tax"]=>
        string(5) "-5.49"
        ["taxes"]=>
        array(1) {
          [0]=>
          array(3) {
            ["id"]=>
            int(1)
            ["total"]=>
            string(7) "-5.4855"
            ["subtotal"]=>
            string(7) "-5.4855"
          }
        }
        ["meta"]=>
        array(0) {
        }
      }
    }
    ["_links"]=>
    array(3) {
      ["self"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(68) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds/581"
        }
      }
      ["collection"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(64) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds"
        }
      }
      ["up"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(56) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578"
        }
      }
    }
  }
  [1]=>
  array(6) {
    ["id"]=>
    int(580)
    ["date_created"]=>
    string(19) "2017-02-28T00:44:26"
    ["amount"]=>
    string(5) "19.56"
    ["reason"]=>
    string(0) ""
    ["line_items"]=>
    array(1) {
      [0]=>
      array(14) {
        ["id"]=>
        int(27)
        ["name"]=>
        string(13) "Woo Single #1"
        ["sku"]=>
        string(10) "woosingle1"
        ["product_id"]=>
        int(93)
        ["variation_id"]=>
        int(0)
        ["quantity"]=>
        int(-2)
        ["tax_class"]=>
        string(0) ""
        ["price"]=>
        string(6) "-15.90"
        ["subtotal"]=>
        string(6) "-15.90"
        ["subtotal_tax"]=>
        string(5) "-3.66"
        ["total"]=>
        string(6) "-15.90"
        ["total_tax"]=>
        string(5) "-3.66"
        ["taxes"]=>
        array(1) {
          [0]=>
          array(3) {
            ["id"]=>
            int(1)
            ["total"]=>
            string(6) "-3.657"
            ["subtotal"]=>
            string(6) "-3.657"
          }
        }
        ["meta"]=>
        array(0) {
        }
      }
    }
    ["_links"]=>
    array(3) {
      ["self"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(68) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds/580"
        }
      }
      ["collection"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(64) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds"
        }
      }
      ["up"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(56) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578"
        }
      }
    }
  }
  [2]=>
  array(6) {
    ["id"]=>
    int(579)
    ["date_created"]=>
    string(19) "2017-02-28T00:44:14"
    ["amount"]=>
    string(4) "9.78"
    ["reason"]=>
    string(0) ""
    ["line_items"]=>
    array(1) {
      [0]=>
      array(14) {
        ["id"]=>
        int(25)
        ["name"]=>
        string(13) "Woo Single #1"
        ["sku"]=>
        string(10) "woosingle1"
        ["product_id"]=>
        int(93)
        ["variation_id"]=>
        int(0)
        ["quantity"]=>
        int(-1)
        ["tax_class"]=>
        string(0) ""
        ["price"]=>
        string(5) "-7.95"
        ["subtotal"]=>
        string(5) "-7.95"
        ["subtotal_tax"]=>
        string(5) "-1.83"
        ["total"]=>
        string(5) "-7.95"
        ["total_tax"]=>
        string(5) "-1.83"
        ["taxes"]=>
        array(1) {
          [0]=>
          array(3) {
            ["id"]=>
            int(1)
            ["total"]=>
            string(7) "-1.8285"
            ["subtotal"]=>
            string(7) "-1.8285"
          }
        }
        ["meta"]=>
        array(0) {
        }
      }
    }
    ["_links"]=>
    array(3) {
      ["self"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(68) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds/579"
        }
      }
      ["collection"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(64) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578/refunds"
        }
      }
      ["up"]=>
      array(1) {
        [0]=>
        array(1) {
          ["href"]=>
          string(56) "https://example-woocommerce.dev/wp-json/wc/v1/orders/578"
        }
      }
    }
  }
  [3]=>
  array(2) {
    ["errors"]=>
    array(1) {
      ["woocommerce_rest_invalid_order_refund_id"]=>
      array(1) {
        [0]=>
        string(24) "Invalid order refund ID."
      }
    }
    ["error_data"]=>
    array(1) {
      ["woocommerce_rest_invalid_order_refund_id"]=>
      int(404)
    }
  }
  [4]=>
  array(2) {
    ["errors"]=>
    array(1) {
      ["woocommerce_rest_invalid_order_refund_id"]=>
      array(1) {
        [0]=>
        string(24) "Invalid order refund ID."
      }
    }
    ["error_data"]=>
    array(1) {
      ["woocommerce_rest_invalid_order_refund_id"]=>
      int(404)
    }
  }
  [5]=>
  array(2) {
    ["errors"]=>
    array(1) {
      ["woocommerce_rest_invalid_order_refund_id"]=>
      array(1) {
        [0]=>
        string(24) "Invalid order refund ID."
      }
    }
    ["error_data"]=>
    array(1) {
      ["woocommerce_rest_invalid_order_refund_id"]=>
      int(404)
    }
  }
  [6]=>
  array(2) {
    ["errors"]=>
    array(1) {
      ["woocommerce_rest_invalid_order_refund_id"]=>
      array(1) {
        [0]=>
        string(24) "Invalid order refund ID."
      }
    }
    ["error_data"]=>
    array(1) {
      ["woocommerce_rest_invalid_order_refund_id"]=>
      int(404)
    }
  }
}
```

`
### WordPress Environment ###

Home URL: https://example-woocommerce.dev
Site URL: https://example-woocommerce.dev
WC Version: 2.6.14
Log Directory Writable: ✔
WP Version: 4.7.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: Apache/2.4.18 (Ubuntu)
PHP Version: 7.0.15-1+deb.sury.org~xenial+1
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 2.6.14
: 
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ❌ The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from http://dev.maxmind.com/geoip/legacy/geolite/ to the path: . Scroll down to \"Downloads\" and download the \"Binary / gzip\" file next to \"GeoLite Country\"

### Active Plugins (2) ###

WooCommerce: by WooThemes – 2.6.14
WordPress Importer: by wordpressdotorg – 0.6.3

### Settings ###

Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2

### API ###

API Enabled: ✔

### WC Pages ###

Shop Base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My Account: #7 - /my-account/

### Taxonomies ###

Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

### Theme ###

Name: Twenty Sixteen
Version: 1.3
Author URL: https://wordpress.org/
Child Theme: ❌ – If you're modifying WooCommerce on a parent theme you didn't build personally
then we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

### Templates ###

Overrides: –
`

**Comment by claudiosanches** [comment]
Have you removed any refund or changed something from the database?

**Comment by luiscoutinh** [comment]
No.

**Comment by claudiosanches** [comment]
Definitely we can't use `return array_filter($results, function($refund) { return !isset($refund['errors']);});` since if something is broke need to be displayed and fixed as possible.
But I need more details from you if it is really an issue with WooCommerce.

**Comment by luiscoutinh** [comment]
Sure,

I don't have change anything in the API side, I'm only doing this in my side only to get this working now, while this issue is not solved.

Tomorrow, I'll have some free time and I'll try find the bug or give you more info about.

**Comment by mikejolley** [comment]
I can see the error here; I'll see if 2.7.x is affected too. Basically, we have this code in 2.6.x:

```
$refund = wc_get_order( $post );

		if ( ! $refund || intval( $refund->post->post_parent ) !== intval( $order->id ) ) {
			return new WP_Error( 'woocommerce_rest_invalid_order_refund_id', __( 'Invalid order refund ID.', 'woocommerce' ), 404 );
		}
```

`wc_get_order` uses the order factory to get order objects. We're passing a WP_POST object.

```
if ( false === $the_order ) {
			$the_order = $post;
		} elseif ( is_numeric( $the_order ) ) {
			$the_order = get_post( $the_order );
		} elseif ( $the_order instanceof WC_Order ) {
			$the_order = get_post( $the_order->id );
		}
```

The order factory does not handle that case.

**Comment by mikejolley** [comment]
Confirmed this is fixed in 2.7+ We have handling for WP_Post objects in the order factory class.

It's unlikely we'll have another 2.6.x release but I'll patch anyway.
